File: aoc/app/d11.hs
{-# LANGUAGE ViewPatterns #-}

import AocLib

import Debug.Trace

import qualified Data.Map as Map
import Data.Map(Map, (!))
import qualified Data.Set as Set
import Data.Set(Set)
import Data.Char
import Data.List
import Data.Maybe
import Data.Either
import Data.Functor
import Data.Function
import Data.Foldable


import Control.Arrow
import Control.Monad.Trans.State.Strict

import Text.ParserCombinators.ReadP ()
import qualified Text.ParserCombinators.ReadP as P


type S = Map Int [Integer]
type C = Map Int Integer

process :: [String] -> [String]
process = fmap (filter (not . (`elem` ",:")))

data M = M { m :: Int, op :: (Integer -> Integer), d :: Integer, tf :: (Int,Int) }

instance Show M where 
  show (M m _ d tf) = "M " ++ show m

parseOp :: [String] -> Integer -> Integer
parseOp ["old", o, n] =
  case n of 
    "old" -> \x -> op o x x
    _ -> \x -> op o x (read n)
  where 
    op "+" = (+)
    op "*" = (*)

parseMonkey :: [String] -> (Int, [Integer], M)
parseMonkey (process -> 
  [
    ints -> [m],
    ints -> stack,
    drop 3 . words -> op,
    ints -> [d],
    ints -> [t],
    ints -> [f]
  ]) = 
    (m, fmap fromIntegral stack, 
     M m (parseOp op) (fromIntegral d) (t,f))

parse inp = (monkeys,state)
  where
    (ms,stacks,monkeys) = unzip3 . fmap parseMonkey . paragraphs . lines $ inp
    state = Map.fromList $ zip ms stacks

step :: (Integer -> Integer) -> S -> M -> S
step post stk mon =
  Map.insert m [] $ Map.unionWith (++) stk stk'
  where 
    (M m _ _ _) = mon

    check :: M -> Integer -> (Int,Integer)
    check (M m op d (t,f)) (post . op -> x) =
      (if x `mod` d == 0 then t else f, x)

    s' = second (:[]) . check mon <$> (stk ! m)
    stk' = Map.fromListWith (flip (++)) s'

solve n post (monkeys,s0) = x*y
  where 
    repMonkeys = init $ concat $ replicate n monkeys

    states = scanl (step post) s0 repMonkeys

    num = length monkeys
    get mon state = state ! m mon

    -- stack processed by each monkey at each step
    stacks = 
      zipWith ($) (cycle (fmap get monkeys)) states
    
    -- number of items for each monkey at each step,
    -- monkey-major order.
    byMonkey = transpose $ chunks num $ fmap length stacks

    counts = fmap sum byMonkey
    x:y:_ = reverse $ sort counts

one = solve 20 (`div` 3)

two x@(monkeys,state) = solve 10000 (`mod` d') x
  where 
    d' = product $ fmap d monkeys

main :: IO ()
main = do
  inp <- parse <$> getContents
  print $ inp
  -- traverse print $ one inp
  print $ one inp
  print $ two inp

File: py/d11.py
import sys 
from .lib import *

from collections import defaultdict

def one():
  d = sys.stdin.read()

  # list of stack
  stacks = defaultdict(lambda: [])
  monkeys = defaultdict(lambda: [None,None,[None,None]])

  ops = {
    '*': lambda x,y: x*y,
    '+': lambda x,y: x+y
  }

  for l in d.split('\n'):
    if not l: continue 
    l = l.split()
    if l[0] == 'Monkey':
      m = int(l[1].strip(':'))
    elif l[0] == 'Starting':
      stacks[m] = [int(x.strip(',')) for x in l[2:]]
    elif l[0] == 'Operation:':
      # print(l)
      o = ops[l[4]]
      if l[5] == 'old':
        l = (lambda o: lambda x: o(x,x))(o)
      else:
        n = int(l[5])
        l = (lambda o,n: (lambda x: o(x,n)))(o,n)
      monkeys[m][0] = l
    elif l[0] == 'Test:':
      d = int(l[3])
      monkeys[m][1] = d 
    elif l[1] == 'true:':
      monkeys[m][2][0] = int(l[-1])
    elif l[1] == 'false:':
      monkeys[m][2][1] = int(l[-1])

  counts = defaultdict(int)

  divv = 1 
  for _,d,_ in monkeys.values():
    divv *= d

  print(stacks)
  print(monkeys)

  def step():
    for m, y in monkeys.items():
      (op, mod, x) = y
      t,f = x
      for x in list(stacks[m]):
        counts[m] += 1
        w = op(x) % divv
        # w = w // 3 
        if w % mod == 0:
          stacks[t].append(w)
        else:
          stacks[f].append(w)
      stacks[m].clear()

  for i in range(10000): 
    # print(i)
    step()

  print(counts)
  x,y = sorted(counts.values(), reverse=True)[:2]
      
  print(x*y)  

if __name__ == '__main__':
  one()
